Tests/Unit/NetworkServer/CupsProtocolMessageProcessorTests.cs

```diff
@@ -110,21 +118,167 @@ public async Task HandleUpdateInfoAsync_Fails_WithInvalidInput(string input, Typ
+                                                CredentialsChecksum,
+                                                string.Empty,
+                                                string.Empty,
+                                                "anotherVersion",
+                                                6789,
+                                                string.Empty);
+            _ = this.basicsStationConfigurationService.Setup(m => m.GetCupsConfigAsync(It.IsAny<StationEui>(), It.IsAny<CancellationToken>()))
+                                                      .Returns(Task.FromResult(cupsTwinInfo));
+            // act
+            var exception = await Assert.ThrowsAsync<InvalidOperationException>(async () => await this.processor.HandleUpdateInfoAsync(httpContext.Object, default));
+
+            Assert.Contains("checksum is not available", exception.Message, StringComparison.OrdinalIgnoreCase);
+        }
+
+        [Theory]
+        [InlineData(0)]
+        [InlineData(long.MaxValue)]
+        public async Task HandleUpdateInfoAsync_Fails_WithNotValidFirmwareLength(long contentLength)
+        {
+            // setup
+            var strictifiedInput = JsonUtil.Strictify(CupsRequestJson);
+            var (httpContext, httpRequest, _) = SetupHttpContextWithRequest(strictifiedInput, null);
+            _ = httpRequest.Setup(r => r.ContentLength).Returns(Encoding.UTF8.GetByteCount(strictifiedInput));
+
+            // setting up the twin in such a way that there is a fw update but no matching checksum
+            var cupsTwinInfo = new CupsTwinInfo(new Uri(CupsUri),
+                                                new Uri(TcUri),
+                                                CredentialsChecksum,
+                                                CredentialsChecksum,
+                                                string.Empty,
+                                                string.Empty,
+                                                "anotherVersion",
+                                                6789,
+                                                string.Empty);
+
+            using var httpContent = new StringContent("firmware");
+            httpContent.Headers.ContentLength = contentLength;
+
+            _ = this.deviceAPIServiceBase.Setup(m => m.FetchStationFirmwareAsync(It.IsAny<StationEui>(), It.IsAny<CancellationToken>()))
+                                         .ReturnsAsync(httpContent);
+            _ = this.basicsStationConfigurationService.Setup(m => m.GetCupsConfigAsync(It.IsAny<StationEui>(), It.IsAny<CancellationToken>()))
+                                                      .Returns(Task.FromResult(cupsTwinInfo));
+            // act
+            var exception = await Assert.ThrowsAsync<InvalidOperationException>(async () => await this.processor.HandleUpdateInfoAsync(httpContext.Object, default));
+            Assert.Contains("firmware", exception.Message, StringComparison.OrdinalIgnoreCase);
```

Comment:
Don't think this is good enough as many of the exception message contain the word firmware. When I checked, you're actually getting the message, “Remote firmware signature generated with a key whose checksum is not available in CUPS request.” I don't think that's what's expected based on the title of the test.